Release r1.1 with v0.1.0-rc.1

XunliYang opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
* subproject management



#### What this PR does / why we need it:

Prepare the release r1.1 with v0.1.0-rc.1 API version:

* Update CHANGELOG.md
* Update API Readiness Checklist
* Update versions in all files

#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #58 



#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.



```
docs

```

XunliYang commented:
> aligned with the upcoming releases r3.2 of Commonalities and ICM

XunliYang commented:
> Created [camaraproject/ReleaseManagement#218](https://github.com/camaraproject/ReleaseManagement/issues/218) for the review by Release Management team.
> 
> Note: as this is a release candidate for Fall25 it needs to be aligned with the upcoming releases r3.2 of Commonalities and ICM.

Thanks herbert, I have modified based on the release r3.2 of Commonalities and ICM. However there is not release r3.2 yet, I can't link to it. So I linked to r3.1 for the moment.

XunliYang commented on CHANGELOG.md:
```diff
@@ -1,22 +1,42 @@
 # Changelog Network Slice Booking API
```
ok, I have removed it.

XunliYang commented on README.md:
```diff
@@ -22,10 +22,11 @@ Repository to describe, develop, document and test the Network Slice Booking API
 <!-- Pre-releases of this sub project are available in https://github.com/camaraproject/§repo_name§/releases -->
```
Thanks, I have applied the suggestion.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
ok, I have added it.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
oh sorry, it's a mistake, I have modified it

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
thanks，I have changed "terminal" to "device" and expand the description "UL" and "DL"

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
Thanks for the suggestion, I'll check and modify the common schemas.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
Thanks for the suggestion, the 5xx error codes are used as their default meaning. We think it is unnecessary to reflect it in the documentation. I will remove the 5xx error codes and supplement the description of schemas.

XunliYang commented on README.md:
Thanks for the suggestions. For line 19, I think the SLA assurance capabilities are reflected in the post request method, which is sending request parameters such as MaxNumofTerminals, DLThroughputPerDevice, ULThroughputPerDevice, DLLatency, and ULLatency to acquire different Qos for different API consumers. The post request method is to predict the SLA can be guaranteed or not for the indicated time and location, and provide resource reserved network slice to guarantee the service QoS.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
Thanks a lot, I have made the modification and expand the description field.

XunliYang commented on documentation/API_documentation/network-slice-booking-API-Readiness-Checklist.md:
Thanks, I have added link to API description https://camaraproject.org/network-slice-booking/, but I'm not sure if it is right. Also I'm wondering how to provide the API description for M4, because I didn't see that it can be edited on this page.

hdamker replied:
@XunliYang: The source of the API description on the website is here on the wiki page: https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/81166982/NetworkSliceBooking+API+description 
And that is the place to edit it for M4, the Marketing team will take it from there and update the website.

XunliYang commented on code/Test_definitions/network-slice-booking-createSession.feature:
Ok, the same questions have been all changed.

tanjadegroot commented on code/Test_definitions/network-slice-booking-createSession.feature:
```diff
@@ -15,11 +15,11 @@ Feature: CAMARA Network Slice Booking API v0.1.0 - Operations for createSession
 Background: Common createSession setup
     Given an environment at "apiRoot"
-    And the resource "/network-slice-booking/vwip/sessions"
+    And the resource "/network-slice-booking/v0.1rc.1/sessions"
```
```suggestion
    And the resource "/network-slice-booking/v0.1rc1/sessions"
```

XunliYang replied:
Thanks Tanja, I have corrected this kind of problem.

tanjadegroot commented on code/Test_definitions/network-slice-booking-deleteSession.feature:
```diff
@@ -8,11 +8,11 @@ Feature: CAMARA Network Slice Booking API v0.1.0 - Operations for deleteSession
 Background: Common deleteSession setup
     Given an environment at "apiRoot"
-    And the resource "/network-slice-booking/vwip/sessions/{sessionId}"
+    And the resource "/network-slice-booking/v0.1rc.1/sessions/{sessionId}"
```
```suggestion
    And the resource "/network-slice-booking/v0.1rc1/sessions/{sessionId}"
```

tanjadegroot commented on code/Test_definitions/network-slice-booking-getSession.feature:
```diff
@@ -8,11 +8,11 @@ Feature: CAMARA Network Slice Booking API v0.1.0 - Operations for getSession
 Background: Common getSession setup
     Given an environment at "apiRoot"
-    And the resource "/network-slice-booking/vwip/sessions/{sessionId}"
+    And the resource "/network-slice-booking/v0.1rc.1/sessions/{sessionId}"
```
```suggestion
    And the resource "/network-slice-booking/v0.1rc1/sessions/{sessionId}"
```

tanjadegroot reviewed (changes requested):
looks great !
a few suggestion for better understanding of the API by non-Telco users.
and some corrections of the version in the URL in the feature files
after that you should be all set.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
```diff
@@ -247,42 +307,43 @@ components:
           format: date-time
+          description: An instant of time, ending of the TimePeriod. If not included, then the period has no ending date.
           example: "2024-06-01T12:00:00Z"
+      required:
```
Thanks, you are right, it should be required. I'll modify it.

XunliYang commented on code/API_definitions/NetworkSliceBooking.yaml:
```diff
@@ -3,10 +3,50 @@ info:
   title: Network Slice Booking
   description: |
     The Network Slice Booking (NSB) API provides programmable interface for developers to reserve a slice resource of a selected area within a period, and manage device access control as needed.
+    This API predict the SLA can be guaranteed or not for the indicated time and location, and provide resource reserved network slice to guarantee the service QoS.
```
Thanks Tanja, these modifications have indeed made the API description much clearer. I will make some changes based on this.

XunliYang commented:
> looks great ! a few suggestion for better understanding of the API by non-Telco users. and some corrections of the version in the URL in the feature files after that you should be all set.

Thanks Tanja, your suggestions are very helpful to us. I think I have completed all the revisions. If there are any further issues, please feel free to bring them up.

tanjadegroot reviewed (dismissed):
One suggested addition to a description.
Many thanks for all the updates. looks great!
/LGTM - OK from Release Management

chinaunicomyangfan reviewed (approved):
LGTM

hdamker reviewed (changes requested):
Another issue:

API Definition: Uses network-slice-booking:sessions:read
Test Files: Reference network-slice-booking:sessions:get

---

And some minor issues:

externalDocs:
  description: Project documentation at CAMARA  # ❌ Should be "Product"

---

TimePeriod:
  properties:
    startDate:
      type: string
      format: date-time
      description: An instant of time, starting of the TimePeriod.  # ❌ Missing RFC 3339 reference
      
Fix: Add RFC 3339 description:

```
description: |
  An instant of time, starting of the TimePeriod. It must follow [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) and must have time zone. Recommended format is yyyy-MM-dd'T'HH:mm:ss.SSSZ
```

---

Most operations use: "Network Slice Booking Sessions"
GET operation uses: "NSB Sessions"
Fix: Use consistent tag name throughout

---

Test File Typos:
"resouces" should be "resources" in multiple test scenarios
"surpported" should be "supported" in examples

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,863 @@
+      description: Attributes required to create a session
+      type: object
+      properties:
+        ServiceTime:
```
property and similar names follow **lowerCamelCase** naming pattern: serviceTime, serviceArea, qosProfile, maxNumOfDevices, downStreamRatePerDevice, etc.

Schema names follow **UpperCamelCase** naming pattern so ideally the "QoSProfile" should be "QosProfile" (with a lower case "s").

these rules help in readability.

These updates could also wait for M4 but would have impact on test API implementations.

These updates will also impact the .feature files, and possibly other e.g. the documentation file(s)

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,863 @@
+      type: object
+      properties:
+        value:
+          description: Quantity of duration
```
```suggestion
          description: Length of duration
```

tanjadegroot reviewed (dismissed):
I would recommend making the changes on the property names to lowerCamelCase as soon as possible to avoid redoing API implementations, and the description fields for M4. 

Once the lowerCamelCase updates are done (I do not put them as blocking the merge)
/LGTM from Release Management - great work by the team !

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -20,9 +20,9 @@ info:
+    * **serviceTime**: is defined by a start time and an end time, which indicates the period during which the network slice will be reserved.
+    * **serviceArea**: can be defined as a circle or a polygon, allowing for flexible geographical coverage. 
+    * **qosProfile**: includes parameters such as maximum number of devices, downstream and upstream rate per device, and packet delay budget.
     * **SessionId**: is a unique identifier for the session, which is returned when a session is created and used to retrieve or delete the session later.
```
Hi, many thanks for doing these updates. I did not list exhaustively all the names to be updated. 

please check for other cases as well, e.g.  
SessionId --> sessionId

thanks !

XunliYang replied:
Apologies for the oversight. We've double-checked and updated all cases. I think I have corrected all in [commit](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/8228a15203b409cf723e052c407cfd66139c0280). Your feedback is greatly appreciated! Let us know if you spot anything else!

XunliYang commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,863 @@
+        - serviceArea
+        - qosProfile
+
+    QosProfile:
```
Thanks Herbert, this indeed makes sense. We have created issue #65 to track it. How about `SliceQosProfile`? Maybe it‘s more brief and also indicates the qosProfile for sclice.

hdamker replied:
`SliceQosProfile` is also fine for me.

XunliYang replied:
@hdamker @tanjadegroot ：
Thanks for your comments, we have corrected naming pattern in [Naming pattern correction](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/8228a15203b409cf723e052c407cfd66139c0280), and modified the `QosProfile` to `SliceQosProfile` in [Modify the name for schema QosProfile to avoid confusions](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/01628c7ff1f39fad54284ec5e4ddc5a4fb9d631d). 
Could you please review the updates again and share your approval or feedback at your earliest convenience? Thank you!

hdamker reviewed (dismissed):
@XunliYang As far as I can see are all review comments from @tanjadegroot addressed which is great. Unfortunately I found with a little help of AI some further issues.

## High priority (should be addressed before release)

Replace 204 usage with appropriate error codes (400/422)

- Issue: HTTP 204 means success and "No Content" and should not have a response body. Using it for failures is non-standard. 
- Fix: Use appropriate 4xx codes (e.g., 422 UNPROCESSABLE_ENTITY or 400 with specific error codes). Correct also the test definition, the tests expect response body for 204, which contradicts HTTP standards.

---

Standardize error message enums

* Issue: Inconsistent Error Message Format
```
message:
  enum:
    - SERVICEAREA NOT SUPPORTED  # Has space
    - RESOURCES INSUFFICIENT
```
Fix: Use consistent format and consider to define as API specific: NETWORK_SLICE_BOOKING.SERVICEAREA_NOT_SUPPORTED 
NETWORK_SLICE_BOOKING.RESOURCES_INSUFFICIENT

Note: the definition as specific API error codes I see as medium priority, if you address it, please check for other API specific error codes, and cf. [CAMARA-API-Design-Guide.md#31-standardized-use-of-camara-error-responses](https://github.com/camaraproject/Commonalities/blob/main/documentation/CAMARA-API-Design-Guide.md#31-standardized-use-of-camara-error-responses)

## Low priority (to be considered after M3)

Add description to operation
* Missing description for POST /sessions

XunliYang commented:
@hdamker : Thanks for pointing this out - great catch! We'd actually considered standardizing these operation error codes and creating custom business logic failure codes before realizing it wasn't the right approach. Really appreciate you flagging this. Here are the changes we've implemented:

##  `error codes`

We have modified in commit [Modify error codes usage with appropriate error codes](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/c3d8ec5cc6929ba3cf10a39b4b001518c3f25c01). The changes include:

- `POST` operation changed error code `204` to `422`, and error message consolidated into a single API-specific definition：`NETWORK_SLICE_BOOKING.RESOURCES_NOT_APPLICABLE`
- `POST` operation changed error code `200` to `201`
- `Delete` operation changed error code `200` to `204`
## `operation description`
We have modified in [Add operation description](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/f09290f8e8b599825cff77ba280eb40274232a85). Further updates will be continuously provided and tracked at #64 .

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,865 @@
+        "404":
+          $ref: "#/components/responses/Generic404"
+        "429":
+          $ref: "#/components/responses/Generic429_requests"
```
```suggestion
          $ref: "#/components/responses/Generic429"
```

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,865 @@
+        "404":
+          $ref: "#/components/responses/Generic404"
+        "429":
+          $ref: "#/components/responses/Generic429_requests"
```
```suggestion
          $ref: "#/components/responses/Generic429"
```

tanjadegroot commented on code/API_definitions/network-slice-booking.yaml:
```diff
@@ -1,831 +1,865 @@
+                code: TOO_MANY_REQUESTS
+                message: Either out of resource quota or reaching rate limiting.
+
+    Generic429_requests:
```
I think this schema can be removed as it is the same as the Generic429. I have proposed above the update of the 2 occurrences of this one to be Generic429 instead.

Also align the **examples** section of the Generi429 error definition with the definitions in CAMARA_common.yaml (see https://github.com/camaraproject/Commonalities/blob/r3.2/artifacts/CAMARA_common.yaml) 

Maybe also check the other error examples to be aligned

tanjadegroot replied:
Hi @XunliYang If you compare with the latest in https://github.com/camaraproject/Commonalities/blob/r3.2/artifacts/CAMARA_common.yaml the Generic429 seems the same as your definition of Generic429_requests - so ideally you would use the name from CAMARA_common one but remove the code the `QUOTA_EXCEEDED` one.

I will verify if that is allowed, but the Guidelines do not say that all error codes need to be supported by the API provider.

For now, let's keep what you have done to meet the M3 deadline

tanjadegroot commented on code/Test_definitions/network-slice-booking-createSession.feature:
```diff
@@ -1,5 +1,5 @@
 @NetworkSliceBooking
-Feature: CAMARA Network Slice Booking API v0.1.0 - Operations for createSession
+Feature: CAMARA Network Slice Booking API v0.1.0-rc.1 - Operations for createSession
```
```suggestion
Feature: CAMARA Network Slice Booking API v0.1.0-rc.1 - Operation createSession
```

tanjadegroot commented on code/Test_definitions/network-slice-booking-createSession.feature:
```diff
@@ -42,27 +42,27 @@ Scenario: Common validations for any success scenario
-  Given starttime, endtime, the request body property "$.ServiceArea" is set to the service area of not supported and the configuration of information of network slicing
+  Given starttime, endtime, the request body property "$.serviceArea" is set to the service area of not supported and the configuration of information of network slicing
   When the request "createSession" is sent
   Then the response status code is 204
```
need to update this to the same error code as done in the API yaml, instead of a success code (204).

same for the other 204 "fail" scenarios below.

tanjadegroot commented on code/Test_definitions/network-slice-booking-createSession.feature:
line 132 and 143: update the messages to the same as in the Generic429 definition (per CAMARA_common.yaml)

tanjadegroot commented on code/Test_definitions/network-slice-booking-deleteSession.feature:
Apply same changes as in comments on the createSession.feature file (if applicable)

tanjadegroot commented on code/Test_definitions/network-slice-booking-getSession.feature:
Apply same changes as in comments on the createSession.feature file (if applicable)

XunliYang commented:
@tanjadegroot: Thanks for your valuable feedback. We've implemented the following modifications accordingly in commit [Modify error codes 200 201 429](https://github.com/camaraproject/NetworkSliceBooking/pull/59/commits/7b7d6b5380429e1658ee1fc9b5e4a60e17c2fc45):

- `Generic429` and `Generic429_request`
Remained  `Generic429_request`, and modified the code message both to align with the definitions in CAMARA_common.yaml.**(To be determinded)**
- `Operations for createSession` to `Operation createSession`
Fixed in all feature files.
- Error code `200` and `201`
Streamlined the response with removing `status` and `result`(#67 ). And added `SessionInfo` to the response of `201`.
- Feature files
The modifications related to error codes have been synchronized across all feature files.

tanjadegroot reviewed (approved):
thanks for the hard work on this API.
It looks great !
/LGTM from Release Management

XunliYang commented:
@hdamker： We believe all requested changes have been addressed. Could you kindly re-review and share your feedback or approval? Thank you for your time!

tanjadegroot commented:
@XunliYang  the review is complete.

I checked on the error code, and indeed you are allowed to remove error code from the list  for the Generic429 to adust it to the API need (as for other optional error codes (see Commonalities API Guide section 3.1 under the tables): 

Error statuses 400, 404, 409, 422, ****: These error statuses SHOULD be documented based on the API design and the functionality involved. Subprojects evaluate the relevance and necessity of including these statuses in API specifications.

so I would still recommend to align the error name to Generic429 instead of an API specific one before merging if you have time, or create an issie in the repo to address it later.

XunliYang commented:
> @XunliYang the review is complete.
> 
> I checked on the error code, and indeed you are allowed to remove error code from the list for the Generic429 to adust it to the API need (as for other optional error codes (see Commonalities API Guide section 3.1 under the tables):
> 
> Error statuses 400, 404, 409, 422, ****: These error statuses SHOULD be documented based on the API design and the functionality involved. Subprojects evaluate the relevance and necessity of including these statuses in API specifications.
> 
> so I would still recommend to align the error name to Generic429 instead of an API specific one before merging if you have time, or create an issie in the repo to address it later.

@tanjadegroot @hdamker Thanks for your suggestions. We will merge this PR to meet the M3 deadline. Also we have created an issue #68 to track it later and modify after this PR. However I see that other APIs also have specific error names such as [QOD](https://github.com/camaraproject/QualityOnDemand/blob/main/code/API_definitions/quality-on-demand.yaml)'s  `CreateSessionBadRequest400` and `GenericExtendSessionDuration400`. We have both error names `Generic429` and `Generic429_request`. Thus I'm wondering if I need to only use  `Generic429` in different operations.

tanjadegroot commented:
> @tanjadegroot @hdamker Thanks for your suggestions. We will merge this PR to meet the M3 deadline. Also we have created an issue #68 to track it later and modify after this PR. However I see that other APIs also have specific error names such as [QOD](https://github.com/camaraproject/QualityOnDemand/blob/main/code/API_definitions/quality-on-demand.yaml)'s `CreateSessionBadRequest400` and `GenericExtendSessionDuration400`. We have both error names `Generic429` and `Generic429_request`. Thus I'm wondering if I need to only use `Generic429` in different operations.

yes, you can make your own error codes as needed, but then they are not "**Generic**429", but it has a specific name for the specific error. In your case, the error is the same as Generic429 with one less opetion in the codes enum, so no need for a new one. 

The QoD API was one of the first APIs defined in CAMARA, and was done before the Commonalities rules for error codes were established. So they may  have some deviations for backward compatibility reasons.

For new APIs, we try to make developers life as easy as possible by having consistent usage in specifications.

Thank you very much for considering this update.
